### ISSUE TITLE
Chapter 2: testing the LinearRegression

### ISSUE THREAD

**Issue by lipsumar** [ageron/handson-ml#182]
I'm enjoying the book very much but I have a question:

In chapter 2, we create a pre-processing pipeline. This pipeline will - among other things - convert the column `ocean_proximity` into a 1-hot encoded matrix.

When using the full dataset, the 1-hot encoding produces a matrix of shape `(n, 5)`, because there are 5 different categories.

But when I try it out on a few instances from the training set:

```python
some_data = housing.iloc[:5]
some_data_labels = housing_labels.iloc[:5]
some_data_prepared = full_pipeline.fit_transform(some_data)
```

Then `some_data_prepared` will have a different shape than `housing_prepared`:
```
Shape housing_prepared: (16512, 16)
Shape some_data_prepared: (5, 14)
```
I think this is because `some_data` happens to contain only 3 different categories. 

I noticed I can fix my issue by doing `some_data_prepared = full_pipeline.transform(some_data)` instead of `some_data_prepared = full_pipeline.fit_transform(some_data)`. Which sort-of makes sense: fitting the LabelBinarizer with `some_data` will tell it that it only has to handle 3 different labels.

Is it because after doing `housing_prepared = full_pipeline.fit_transform(housing)`, `full_pipeline` will "remember" how it was first `fit`ted ?

**Comment by ageron** [comment]
Hi @lipsumar ,
Thanks for your kind words! My answer will be quite short because your intuition in the last sentence is correct. Indeed, you should only call `fit()` on the training set (or `fit_transform()`)  and never on the validation set, or test set or new instances.  When you call `fit()`, some parameters are learned, such as the number of categories and how they should be mapped to one-hot vectors.  You don't want to learn these parameters on new instances.
I hope this helps!

**Comment by lipsumar** [comment]
Thanks for your answer, this makes sense.
